providers/gitlab.go

```diff
@@ -305,11 +305,24 @@ func (p *GitLabProvider) EnrichSession(ctx context.Context, s *sessions.SessionS
 }
 
 // addGroupsToSession projects into session.Groups
-func (p *GitLabProvider) addGroupsToSession(ctx context.Context, s *sessions.SessionState) {
-	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
-	for _, group := range p.Groups {
-		s.Groups = append(s.Groups, fmt.Sprintf("group:%s", group))
+func (p *GitLabProvider) addGroupsToSession(ctx context.Context, userInfo *gitlabUserInfo, s *sessions.SessionState) {
+	if len(p.Groups) == 0 {
+		return
+	}
+
+	// Collect user group memberships
```

Comment:
This should all be happening here in `Authorize`. That way we decouple authentication from authorization. Since we validate the group memberships in every request in `Authorize`, it allows admins to alter the authorized groups to disallow previous groups that are already authenticated and accessing (rather than only checking membership at login time): https://github.com/oauth2-proxy/oauth2-proxy/blob/f3209a40e16defdab442ef87765e837d90ec4f9a/providers/provider_default.go#L110